File: packages/cli/hieudoanm/cobra.go/cmd/semver/semver.go
package semver

import (
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"

	"github.com/spf13/cobra"
)

var semverJSON bool

type version struct {
	major, minor, patch int
	prerelease          string
}

func parseVersion(s string) (version, error) {
	s = strings.TrimPrefix(s, "v")
	parts := strings.SplitN(s, ".", 3)
	if len(parts) < 3 {
		return version{}, fmt.Errorf("invalid semver: %s (need major.minor.patch)", s)
	}
	major, err := strconv.Atoi(parts[0])
	if err != nil {
		return version{}, fmt.Errorf("invalid major version: %s", parts[0])
	}
	minor, err := strconv.Atoi(parts[1])
	if err != nil {
		return version{}, fmt.Errorf("invalid minor version: %s", parts[1])
	}

	patchStr := parts[2]
	prerelease := ""
	if idx := strings.Index(patchStr, "-"); idx >= 0 {
		prerelease = patchStr[idx+1:]
		patchStr = patchStr[:idx]
	}

	patch, err := strconv.Atoi(patchStr)
	if err != nil {
		return version{}, fmt.Errorf("invalid patch version: %s", patchStr)
	}
	return version{major, minor, patch, prerelease}, nil
}

func (v version) String() string {
	s := fmt.Sprintf("%d.%d.%d", v.major, v.minor, v.patch)
	if v.prerelease != "" {
		s += "-" + v.prerelease
	}
	return s
}

func (a version) compare(b version) int {
	if a.major != b.major {
		return a.major - b.major
	}
	if a.minor != b.minor {
		return a.minor - b.minor
	}
	if a.patch != b.patch {
		return a.patch - b.patch
	}
	if a.prerelease != b.prerelease {
		if a.prerelease == "" {
			return 1
		}
		if b.prerelease == "" {
			return -1
		}
		if a.prerelease < b.prerelease {
			return -1
		}
		return 1
	}
	return 0
}

func (v version) bump(part string) version {
	switch part {
	case "major":
		return version{v.major + 1, 0, 0, ""}
	case "minor":
		return version{v.major, v.minor + 1, 0, ""}
	case "patch":
		return version{v.major, v.minor, v.patch + 1, ""}
	default:
		return v
	}
}

func NewCommand() *cobra.Command {
	var bumpPart, prerelease, rangeExpr string

	cmd := &cobra.Command{
		Use:   "semver <command> [versions...]",
		Short: "Parse, compare, sort, and bump semver strings",
		Long:  `Tools for working with semantic version strings (major.minor.patch).`,
		Example: `  semver validate 1.2.3
  semver compare 1.0.0 2.0.0
  semver sort 1.2.0 2.0.0 1.10.0
  semver --bump minor 1.2.3
  semver --bump patch --prerelease alpha 1.2.3
  semver --range ">=1.0.0 <2.0.0" 1.5.0`,
		Args: cobra.MinimumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if bumpPart != "" {
				v, err := parseVersion(args[0])
				if err != nil {
					return err
				}
				result := v.bump(bumpPart)
				if prerelease != "" {
					result.prerelease = prerelease
				}
				if semverJSON {
					out, _ := json.MarshalIndent(map[string]interface{}{
						"input":     args[0],
						"result":    result.String(),
						"bump":      bumpPart,
						"prerelease": prerelease,
					}, "", "  ")
					fmt.Println(string(out))
				} else {
					fmt.Println(result)
				}
				return nil
			}

			if rangeExpr != "" {
				if len(args) < 1 {
					return fmt.Errorf("need a version to check against range")
				}
				v, err := parseVersion(args[0])
				if err != nil {
					return err
				}
				matches, err := checkRange(v, rangeExpr)
				if err != nil {
					return err
				}
				if semverJSON {
					out, _ := json.MarshalIndent(map[string]interface{}{
						"version": args[0],
						"range":   rangeExpr,
						"matches": matches,
					}, "", "  ")
					fmt.Println(string(out))
				} else {
					if matches {
						fmt.Printf("%s matches range %s\n", v, rangeExpr)
					} else {
						fmt.Printf("%s does NOT match range %s\n", v, rangeExpr)
					}
				}
				return nil
			}

			if len(args) < 2 {
				return fmt.Errorf("need at least 2 arguments for validate|compare|sort")
			}

			action := args[0]
			versions := args[1:]

			switch action {
			case "validate":
				if semverJSON {
					results := make([]map[string]interface{}, 0)
					for _, s := range versions {
						_, err := parseVersion(s)
						valid := err == nil
						entry := map[string]interface{}{
							"version": s,
							"valid":   valid,
						}
						if err != nil {
							entry["error"] = err.Error()
						}
						results = append(results, entry)
					}
					out, _ := json.MarshalIndent(results, "", "  ")
					fmt.Println(string(out))
				} else {
					for _, s := range versions {
						_, err := parseVersion(s)
						if err != nil {
							fmt.Printf("%s: invalid (%v)\n", s, err)
						} else {
							fmt.Printf("%s: valid\n", s)
						}
					}
				}

			case "compare":
				if len(versions) < 2 {
					return fmt.Errorf("compare needs 2 versions")
				}
				a, err := parseVersion(versions[0])
				if err != nil {
					return err
				}
				b, err := parseVersion(versions[1])
				if err != nil {
					return err
				}
				cmp := a.compare(b)
				var rel string
				switch {
				case cmp < 0:
					rel = "<"
				case cmp > 0:
					rel = ">"
				default:
					rel = "=="
				}
				if semverJSON {
					out, _ := json.MarshalIndent(map[string]interface{}{
						"a":        a.String(),
						"b":        b.String(),
						"relation": rel,
					}, "", "  ")
					fmt.Println(string(out))
				} else {
					fmt.Printf("%s %s %s\n", a, rel, b)
				}

			case "sort":
				parsed := make([]version, len(versions))
				for i, s := range versions {
					v, err := parseVersion(s)
					if err != nil {
						return fmt.Errorf("%s: %w", s, err)
					}
					parsed[i] = v
				}
				sort.Slice(parsed, func(i, j int) bool {
					return parsed[i].compare(parsed[j]) < 0
				})
				if semverJSON {
					sorted := make([]string, len(parsed))
					for i, v := range parsed {
						sorted[i] = v.String()
					}
					out, _ := json.MarshalIndent(map[string]interface{}{
						"sorted": sorted,
					}, "", "  ")
					fmt.Println(string(out))
				} else {
					for _, v := range parsed {
						fmt.Println(v)
					}
				}

			default:
				if _, err := parseVersion(action); err == nil && len(versions) >= 0 {
					v, _ := parseVersion(action)
					result := v.bump(bumpPart)
					if semverJSON {
						out, _ := json.MarshalIndent(map[string]interface{}{
							"input":  action,
							"result": result.String(),
							"bump":   bumpPart,
						}, "", "  ")
						fmt.Println(string(out))
					} else {
						fmt.Println(result)
					}
				} else {
					return fmt.Errorf("unknown action: %s (use validate, compare, sort)", action)
				}
			}
			return nil
		},
	}

	cmd.Flags().StringVar(&bumpPart, "bump", "", "Bump version part: major, minor, patch")
	cmd.Flags().StringVar(&prerelease, "prerelease", "", "Set prerelease label after bump")
	cmd.Flags().StringVar(&rangeExpr, "range", "", "Check if version matches a range (e.g. '>=1.0.0 <2.0.0')")
	cmd.Flags().BoolVar(&semverJSON, "json", false, "Output in JSON format")
	return cmd
}

func checkRange(v version, rangeExpr string) (bool, error) {
	parts := strings.Fields(rangeExpr)
	if len(parts) == 0 {
		return false, fmt.Errorf("empty range expression")
	}

	for i := 0; i < len(parts); i += 2 {
		op := parts[i]
		if i+1 >= len(parts) {
			return false, fmt.Errorf("incomplete range: missing version after %s", op)
		}
		ver, err := parseVersion(parts[i+1])
		if err != nil {
			return false, fmt.Errorf("invalid version in range: %s", parts[i+1])
		}

		switch op {
		case ">":
			if v.compare(ver) <= 0 {
				return false, nil
			}
		case ">=":
			if v.compare(ver) < 0 {
				return false, nil
			}
		case "<":
			if v.compare(ver) >= 0 {
				return false, nil
			}
		case "<=":
			if v.compare(ver) > 0 {
				return false, nil
			}
		case "=", "==":
			if v.compare(ver) != 0 {
				return false, nil
			}
		default:
			return false, fmt.Errorf("unknown operator: %s", op)
		}
	}

	return true, nil
}

File: packages/cli/hieudoanm/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/Semver.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.required
import com.google.gson.GsonBuilder

data class SemverVersion(val major: Int, val minor: Int, val patch: Int, val prerelease: String = "") {
    fun compare(other: SemverVersion): Int {
        major.compareTo(other.major).let { if (it != 0) return it }
        minor.compareTo(other.minor).let { if (it != 0) return it }
        patch.compareTo(other.patch).let { if (it != 0) return it }
        if (prerelease != other.prerelease) {
            if (prerelease.isEmpty()) return 1
            if (other.prerelease.isEmpty()) return -1
            return prerelease.compareTo(other.prerelease)
        }
        return 0
    }

    fun bump(part: String): SemverVersion = when (part) {
        "major" -> SemverVersion(major + 1, 0, 0)
        "minor" -> SemverVersion(major, minor + 1, 0)
        "patch" -> SemverVersion(major, minor, patch + 1)
        else -> this
    }

    override fun toString(): String {
        val s = "$major.$minor.$patch"
        return if (prerelease.isNotEmpty()) "$s-$prerelease" else s
    }
}

fun parseSemver(s: String): SemverVersion {
    var sv = s.trimStart('v')
    val parts = sv.split(".", limit = 3)
    require(parts.size >= 3) { "invalid semver: $s (need major.minor.patch)" }
    val major = parts[0].toIntOrNull() ?: throw IllegalArgumentException("invalid major version: ${parts[0]}")
    val minor = parts[1].toIntOrNull() ?: throw IllegalArgumentException("invalid minor version: ${parts[1]}")
    var patchStr = parts[2]
    var prerelease = ""
    val idx = patchStr.indexOf("-")
    if (idx >= 0) {
        prerelease = patchStr.substring(idx + 1)
        patchStr = patchStr.substring(0, idx)
    }
    val patch = patchStr.toIntOrNull() ?: throw IllegalArgumentException("invalid patch version: $patchStr")
    return SemverVersion(major, minor, patch, prerelease)
}

fun checkRange(v: SemverVersion, rangeExpr: String): Boolean {
    val parts = rangeExpr.split("\\s+".toRegex())
    require(parts.isNotEmpty()) { "empty range expression" }
    var i = 0
    while (i < parts.size) {
        val op = parts[i]
        require(i + 1 < parts.size) { "incomplete range: missing version after $op" }
        val ver = parseSemver(parts[i + 1])
        val cmp = v.compare(ver)
        val matches = when (op) {
            ">" -> cmp > 0
            ">=" -> cmp >= 0
            "<" -> cmp < 0
            "<=" -> cmp <= 0
            "=", "==" -> cmp == 0
            else -> throw IllegalArgumentException("unknown operator: $op")
        }
        if (!matches) return false
        i += 2
    }
    return true
}

class SemverCommand : CliktCommand(name = "semver", help = "Parse, compare, sort, and bump semver strings") {
    init {
        subcommands(SemverBump(), SemverCompare(), SemverRange(), SemverSort(), SemverValidate())
    }
    override fun run() = Unit
}

class SemverBump : CliktCommand(name = "bump", help = "Bump a semver version") {
    private val ver by option("--version", help = "Version to bump").required()
    private val major by option("--major", help = "Bump major version").flag()
    private val minor by option("--minor", help = "Bump minor version").flag()
    private val patch by option("--patch", help = "Bump patch version").flag()
    private val pre by option("--pre", help = "Set prerelease label").default("")
    private val build by option("--build", help = "Set build metadata").default("")
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val v = parseSemver(ver)
        val part = when {
            major -> "major"
            minor -> "minor"
            patch -> "patch"
            else -> "patch"
        }
        var result = v.bump(part)
        if (pre.isNotEmpty()) result = result.copy(prerelease = pre)
        val extra = if (build.isNotEmpty()) "+$build" else ""
        val outStr = result.toString() + extra
        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("input" to ver, "result" to outStr, "bump" to part, "prerelease" to pre)))
        } else {
            echo(outStr)
        }
    }
}

class SemverCompare : CliktCommand(name = "compare", help = "Compare two semver strings") {
    private val aVer by option("--a", help = "First version").required()
    private val bVer by option("--b", help = "Second version").required()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val a = parseSemver(aVer)
        val b = parseSemver(bVer)
        val cmp = a.compare(b)
        val rel = when {
            cmp < 0 -> "<"
            cmp > 0 -> ">"
            else -> "=="
        }
        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("a" to a.toString(), "b" to b.toString(), "relation" to rel)))
        } else {
            echo("${a} $rel ${b}")
        }
    }
}

class SemverRange : CliktCommand(name = "range", help = "Check if a version satisfies a range constraint") {
    private val ver by option("--version", help = "Version to check").required()
    private val rangeExpr by option("--range", help = "Range constraint (e.g. '>=1.0.0 <2.0.0')").required()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val v = parseSemver(ver)
        val matches = checkRange(v, rangeExpr)
        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("version" to ver, "range" to rangeExpr, "matches" to matches)))
        } else {
            if (matches) echo("$v matches range $rangeExpr")
            else echo("$v does NOT match range $rangeExpr")
        }
    }
}

class SemverSort : CliktCommand(name = "sort", help = "Sort semver strings") {
    private val versions by option("--versions", help = "Comma-separated versions").required()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val versionList = versions.split(",").map { it.trim() }
        val parsed = versionList.map { parseSemver(it) }
        val sorted = parsed.sortedWith { a, b -> a.compare(b) }
        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("sorted" to sorted.map { it.toString() })))
        } else {
            sorted.forEach { echo(it.toString()) }
        }
    }
}

class SemverValidate : CliktCommand(name = "validate", help = "Validate semver strings") {
    private val versions by option("--versions", help = "Comma-separated versions").required()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val versionList = versions.split(",").map { it.trim() }
        if (json) {
            val results = versionList.map { s ->
                val valid = try { parseSemver(s); true } catch (e: Exception) { false }
                mapOf("version" to s, "valid" to valid)
            }
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(results))
        } else {
            versionList.forEach { s ->
                try {
                    parseSemver(s)
                    echo("$s: valid")
                } catch (e: Exception) {
                    echo("$s: invalid (${e.message})")
                }
            }
        }
    }
}
